### PR TITLE
Add insecure flag to allow for empty client secrets

### PR DESCRIPTION
**Pull request by paulczar**
UAA as used by Enterprise PKS and Cloud Foundry does not have a client-secret set
for the OIDC client. This flag allows us to omit the secret.

Signed-off-by: Paul Czarkowski <username.taken@gmail.com>

## Description

It's unusual but permissable to have an empty client secret.  This PR adds a command line flag to allow the user to omit the `client-secret` flag.

## Motivation and Context

This is needed to get oauth_proxy to authenticate against a UAA server, specifically to allow [kubeapps](https://github.com/kubeapps/kubeapps) to authenticate a user via oauth2-proxy and gain their access to a PKS kubernetes cluster

## How Has This Been Tested?

Can Build and run in kubernetes,  kubeapp protected by oauth2-proxy can authenticate appropriately.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -3,6 +3,9 @@ FROM golang:1.13-buster AS builder
 # Download tools
 RUN curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $(go env GOPATH)/bin v1.17.1
 
+RUN wget -O- -nv https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s v1.23.8
```
Comment (JoelSpeed) [inline Dockerfile]: Why is this adding a second copy of `golangci-lint`?

File: options.go [hunk 2]
```diff
@@ -89,19 +89,20 @@ type Options struct {
...
-	ProtectedResource                string `flag:"resource" cfg:"resource" env:"OAUTH2_PROXY_RESOURCE"`
-	ValidateURL                      string `flag:"validate-url" cfg:"validate_url" env:"OAUTH2_PROXY_VALIDATE_URL"`
-	Scope                            string `flag:"scope" cfg:"scope" env:"OAUTH2_PROXY_SCOPE"`
-	ApprovalPrompt                   string `flag:"approval-prompt" cfg:"approval_prompt" env:"OAUTH2_PROXY_APPROVAL_PROMPT"`
+	Provider                           string `flag:"provider" cfg:"provider" env:"OAUTH2_PROXY_PROVIDER"`
+	ProviderName                       string `flag:"provider-display-name" cfg:"provider_display_name" env:"OAUTH2_PROXY_PROVIDER_DISPLAY_NAME"`
+	OIDCIssuerURL                      string `flag:"oidc-issuer-url" cfg:"oidc_issuer_url" env:"OAUTH2_PROXY_OIDC_ISSUER_URL"`
+	InsecureOIDCAllowUnverifiedEmail   bool   `flag:"insecure-oidc-allow-unverified-email" cfg:"insecure_oidc_allow_unverified_email" env:"OAUTH2_PROXY_INSECURE_OIDC_ALLOW_UNVERIFIED_EMAIL"`
+	InsecureOIDCAllowEmptyClientSecret bool   `flag:"insecure-oidc-allow-empty-client-secret" cfg:"insecure_oidc_allow_unverified_email" env:"OAUTH2_PROXY_INSECURE_OIDC_ALLOW_EMPTY_CLIENT_SECRET"`
```
Comment (JoelSpeed) [inline options.go]:
```suggestion
	InsecureOIDCAllowEmptyClientSecret bool   `flag:"insecure-oidc-allow-empty-client-secret" cfg:"insecure_oidc_allow_empty_client_secret" env:"OAUTH2_PROXY_INSECURE_OIDC_ALLOW_EMPTY_CLIENT_SECRET"`
```

File: Dockerfile (hunk 1 above)
Comment (paulczar) [reply Dockerfile]: oh woops I'll rip that out. I had some initial issues with the image and the included golangci-lint.

File: options.go (hunk 2 above)
Comment (paulczar) [reply options.go]: good catch!

**Comment by JoelSpeed** [comment]
@paulczar I'm struggling to find reference to where UAA specifies that it doesn't need a client secret, could you please point me to the documentation that clarifies this for me?

**Comment by paulczar** [comment]
Hi,

Several Pivotal products ship with using an empty client-secret to auth via OIDC with UAA such as Cloud Foundry and Pivotal Container Service. Here's an example from my kube config:

```
- name: pksuser
  user:
    auth-provider:
      config:
        client-id: pks_cluster_client
        client-secret: ""

```
I'm hunting for UAA docs that specify its ok, but cannot find right now. However I can link you to a similar change made in the gangway project to accommodate:

https://github.com/heptiolabs/gangway/pull/84

You'll also note I linked to the RFC in that PR (https://tools.ietf.org/html/rfc6749#section-2.3.1) which suggests that its okay to omit the client secret if its an empty string:

```
   client_secret
         REQUIRED.  The client secret.  The client MAY omit the
         parameter if the client secret is an empty string.
```

**Comment by paulczar** [comment]
heh looks like my attempt to merge conflicts went poorly.

**Pull request closed without merging**